Участвовать в событии

Rhanan1233 opened the pull request.

Rhanan1233 committed:
Участвовать в событии

senty8 commented:
1. Название ветки почему-то add skills, хотя логичнее сделать что-то связанное с event (например participate-event). Но учитывая, что мы на дейли обсуждали, можно пока и так оставить, если не хочешь разбираться как менять название.
2. Лучше избегать кириллицы в коде (сообщения в exception), на крайняк лучше перевести через переводчик, ибо могут возникать ошибки из-за кодировок на разных устройствах.

Rhanan1233 committed:
Участвовать в событии

Rhanan1233 committed:
подправил метод регистрации

Rhanan1233 committed:
Убрал аннотацию lombok
подправил метод регистрации

Rhanan1233 committed:
Создал mapper, Поменял некоторые аннотации lombok, убрал логику в классе controller

senty8 commented:
Есть незначительное замечание по сокращению кода, также надо сделать тесты, в остальном все отлично, отправляю на ревью.

Rhanan1233 committed:
Добавил нужные мне аннотации,
Сократил код в Service добавив
private метод на проверку
исключений, написал Unit тесты.

SnurnicynKonstantin commented on src/main/java/school/faang/user_service/service/event/EventParticipationService.java:
```diff
@@ -0,0 +1,54 @@
+    public List<UserDto> getParticipant(long eventId) {
+        List<User> users = eventParticipationRepository.findAllParticipantsByEventId(eventId);
+        List<UserDto> dto = userMapper.usersToUserDtos(users);
+        return dto;
```
Можно сразу `return userMapper.usersToUserDtos(users);`

SnurnicynKonstantin commented on src/main/java/school/faang/user_service/service/event/EventParticipationService.java:
```diff
@@ -0,0 +1,54 @@
+                .unregister(eventId, userId);
+    }
+
+    private boolean isParticipantRegistered(long eventId, long userId) {
```
Сначала идут публичные методы, потом приватные. Это общие соглашения, которые используются в java. Перенеси этот метод в конец файла плиз.

SnurnicynKonstantin commented on src/main/java/school/faang/user_service/service/event/EventParticipationService.java:
```diff
@@ -0,0 +1,54 @@
+    public void registerParticipant(long eventId, long userId) {
+
+        if (isParticipantRegistered(eventId, userId)) {
+            throw new IllegalArgumentException("The user is already a participant in the event! ");
```
Это исключение и в следующем методе можно вынести в сам метод `isParticipantRegistered`. Иначе дублируется один и тот же код.

Тогда `isParticipantRegistered` может выглядеть как:
```
eventParticipationRepository.findAllParticipantsByEventId(eventId)
        .stream()
        .filter(user -> user.getId() == userId)
        .findFirst()
        .orElseThrow(() -> throw new IllegalArgumentException("The user is already a participant in the event! "));
```

Rhanan1233 replied:
А как же текст исключения?

SnurnicynKonstantin commented on src/test/java/school/faang/user_service/service/EventParticipationServiceTest.java:
```diff
@@ -0,0 +1,133 @@
+    private EventParticipationRepository eventParticipationRepository;
+
+    @Spy
+    private UserMapperImpl userMapper;
```
Маппер не нужно мокать. От должен по честному через себя пропускать объекты и трансформировать.

В юнит-тестах моки используются для замены реальных зависимостей тестируемого класса или метода на их "поддельные" реализации. Это позволяет изолировать тестируемый код и исключить влияние внешних факторов, таких как базы данных, сторонние API, файловая система или сложные вычисления.

Мокирование мапперов приводит к замене их поведения заглушками (when(...).thenReturn(...)), что скрывает ошибки в самой логике маппинга. Например, если в маппере произошла ошибка (добавилось поле или удалилось), тесты не обнаружат её, так как мок полностью игнорирует оригинальную реализацию.

Посмотри в этом примере как нужно объявлять сервис - https://github.com/CorporationX/user_service/blob/kraken-master-stream8/src/test/java/school/faang/user_service/service/SubscriptionServiceTest.java#L63

SnurnicynKonstantin reviewed (changes requested):
Все супер, оставил не большие замечания
Плюс сделай rebase на kraken-master-stream8. Иначе сейчас конфликты. Обрати внимание, что UserDto и UserMapper уже созданы в мастере. Тебе надо их переиспользовать, добавить свое если надо.

Rhanan1233 committed:
Добавил объект UserMapper в тест для корректной работы, исправил недочёты в классе Service

Rhanan1233 committed:
Merge branch 'kraken-master-stream8' of https://github.com/CorporationX/user_service into kraken-master-stream8

# Conflicts:
#	src/main/java/school/faang/user_service/dto/UserDto.java
#	src/main/java/school/faang/user_service/mapper/UserMapper.java
